### PR TITLE
Make D2 dates related functions SqlBuilder aware (DHIS-16705)

### PR DESCRIPTION
**Pull request by luciano-fiandesio**
## Summary  
This PR enhances the integration of the `SqlBuilder` with the Antlr parser context and updates D2 functions to utilize the newly introduced `dateDifference` function.

## Changes  
- Introduced a `dateDifference` function within the `SqlBuilder` for improved date manipulation.
- Updated `D2*Between` functions to incorporate the `SqlBuilder` for uniform processing.
- Added unit tests for all modified expression classes.

### CODE REVIEW

File: dhis-2/dhis-support/dhis-support-sql/src/main/java/org/hisp/dhis/db/sql/SqlBuilder.java [hunk 1]
```diff
@@ -422,4 +443,12 @@ public interface SqlBuilder {
    * @return a drop catalog if exists statement.
    */
   String dropCatalogIfExists();
+
+  enum DatePart {
```
Comment (maikelarabori) [inline dhis-2/dhis-support/dhis-support-sql/src/main/java/org/hisp/dhis/db/sql/SqlBuilder.java:447]: Call it DateUnit, maybe?

**Review by maikelarabori — APPROVED** [review]
Looks good 👍 
Minor suggestion added.
Thx!

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java [hunk 2]
```diff
@@ -126,7 +126,7 @@ public DefaultProgramIndicatorService(
     this.analyticsSqlCache = cacheProvider.createAnalyticsSqlCache();
     this.sqlBuilder = sqlBuilder;
 
-    this.programIndicatorItems = new ExpressionMapBuilder(sqlBuilder).getExpressionItemMap();
+    this.programIndicatorItems = new ExpressionMapBuilder().getExpressionItemMap();
```
Comment (gnespolino) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java:129]: I’m curious — was there a particular reason we opted not to use dependency injection in this case?

**Review by gnespolino — APPROVED** [review]
Good job, just a minor comment

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java (hunk 2 above)
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java:129]: I think it's OK. In `DefaultExpressionService` these maps are still generated in the service (for all the expression types that are not program indicators), so this is not consistent with that. But all this ANLR code will be ripped out when we transition to Jan B's new parser. (Did you know about that?) I'm OK with keeping it as a separate class or putting it back where it was.

**Review by jimgrace — APPROVED** [review]
Some code will also need to change at some point under `org.hisp.dhis.parser.expression` if Doris follows MySQL syntax rather than Postgres. Classes I've noticed in a quick review that will need changing are:
`FunctionContains`
`FunctionContainsItems`
`OperatorMathPower`

**Pull request merged**